### PR TITLE
Additional functions to create filters.

### LINKED ISSUE #30: Design wrappers for filter expressions
At the end of this task we will have classes (and standalone functions) to create filter expressions.
The first step is to write a design doc for how they would work, and break down the work into smaller issues.

### PR DESCRIPTION
**Pull request by coryan**
This is another step to fix #30.  It introduces more
functions to create filters, though these are less
common ones.

### CODE REVIEW

File: bigtable/client/filters.h [hunk 1]
```diff
@@ -71,8 +74,12 @@ class Filter {
...
   }
 
-  /// Create a filter that accepts only the last @a n values.
-  static Filter Latest(int n) {
+  /**
+   * Return a filter that accepts only the last @p n values of each column.
+   *
+   * TODO(#84) - document what is the effect of n <= 0
+   */
+  static Filter Latest(std::int32_t n) {
```
Comment (mbrukman) [inline bigtable/client/filters.h]: Can we use an unsigned int here, or is int32 the correct type to pass down the call stack?

File: bigtable/client/filters.h [hunk 2]
```diff
@@ -196,8 +195,9 @@ class Filter {
    * timestamp.
    *
    * TODO(#82) - check the documentation around ordering of columns.
+   * TODO(#84) - document what is the effect of n <= 0
    */
-  static Filter CellsRowLimit(int n) {
+  static Filter CellsRowLimit(std::int32_t n) {
```
Comment (mbrukman) [inline bigtable/client/filters.h]: Same comment here re: unsigned, if possible.

File: bigtable/client/filters.h [hunk 3]
```diff
@@ -214,26 +214,158 @@ class Filter {
    * timestamp.
    *
    * TODO(#82) - check the documentation around ordering of columns.
+   * TODO(#84) - document what is the effect of n <= 0
    */
-  static Filter CellsRowOffset(int n) {
+  static Filter CellsRowOffset(std::int32_t n) {
```
Comment (mbrukman) [inline bigtable/client/filters.h]: Same comment here re: unsigned, if possible.

File: bigtable/client/filters.h [hunk 4]
```diff
@@ -214,26 +214,158 @@ class Filter {
...
     return tmp;
   }
 
+  /**
+   * Return a filter that samples rows with a given probability.
+   *
+   * TODO(#84) - decide what happens if the probability is out of range.
+   *
+   * @param probability the probability that any row will be selected.  It
+   *     must be in the [0.0, 1.0] range.
```
Comment (mbrukman) [inline bigtable/client/filters.h]: > [...] must be in the range [0.0, 1.0].

but you can also simplify it to:

> [...] must be in the range [0, 1].

File: bigtable/client/filters.h [hunk 5]
```diff
@@ -214,26 +214,158 @@ class Filter {
...
+   * @name Less common range filters.
+   *
+   * Cloud Bigtable range filters can include or exclude the limits of the
+   * range.  In most cases applications use [start,end) ranges, and the
+   * ValueRange() and ColumnRange() functions are offered to support the
+   * common case.  For the less common cases where the application needs
+   * different ranges the following functions are available.
+   */
+  /**
+   * Return a filter that accepts values in the [@p start, @p end) range.
```
Comment (mbrukman) [inline bigtable/client/filters.h]: > [...] `values in the range [@p start, @p end).`

File: bigtable/client/filters.h [hunk 6]
```diff
@@ -214,26 +214,158 @@ class Filter {
...
+  static Filter ValueRangeLeftOpen(std::string start, std::string end) {
+    Filter tmp;
+    auto& range = *tmp.filter_.mutable_value_range_filter();
+    range.set_start_value_open(std::move(start));
+    range.set_end_value_closed(std::move(end));
+    return tmp;
+  }
+
+  /**
+   * Return a filter that accepts values in the [@p start, @p end] range.
```
Comment (mbrukman) [inline bigtable/client/filters.h]: > [...] `values in the range [@p start, @p end].`

File: bigtable/client/filters.h [hunk 7]
```diff
@@ -214,26 +214,158 @@ class Filter {
...
+  static Filter ValueRangeRightOpen(std::string start, std::string end) {
+    Filter tmp;
+    auto& range = *tmp.filter_.mutable_value_range_filter();
+    range.set_start_value_closed(std::move(start));
+    range.set_end_value_open(std::move(end));
+    return tmp;
+  }
+
+  /**
+   * Return a filter that accepts values in the [@p start, @p end] range.
```
Comment (mbrukman) [inline bigtable/client/filters.h]: > [...] `values in the range [@p start, @p end].`

File: bigtable/client/filters.h [hunk 8]
```diff
@@ -214,26 +214,158 @@ class Filter {
...
+  static Filter ValueRangeClosed(std::string start, std::string end) {
+    Filter tmp;
+    auto& range = *tmp.filter_.mutable_value_range_filter();
+    range.set_start_value_closed(std::move(start));
+    range.set_end_value_closed(std::move(end));
+    return tmp;
+  }
+
+  /**
+   * Return a filter that accepts values in the (@p start, @p end) range.
```
Comment (mbrukman) [inline bigtable/client/filters.h]: Same as above.

File: bigtable/client/filters.h [hunk 9]
```diff
@@ -214,26 +214,158 @@ class Filter {
...
+  static Filter ValueRangeOpen(std::string start, std::string end) {
+    Filter tmp;
+    auto& range = *tmp.filter_.mutable_value_range_filter();
+    range.set_start_value_open(std::move(start));
+    range.set_end_value_open(std::move(end));
+    return tmp;
+  }
+
+  /**
+   * Return a filter that accepts columns in the [@p start, @p end) range
```
Comment (mbrukman) [inline bigtable/client/filters.h]: Same as above.

File: bigtable/client/filters.h [hunk 10]
```diff
@@ -214,26 +214,158 @@ class Filter {
...
+    Filter tmp;
+    auto& range = *tmp.filter_.mutable_column_range_filter();
+    range.set_family_name(std::move(column_family));
+    range.set_start_qualifier_closed(std::move(start));
+    range.set_end_qualifier_open(std::move(end));
+    return tmp;
+  }
+
+  /**
+   * Return a filter that accepts columns in the (@p start, @p end] range
```
Comment (mbrukman) [inline bigtable/client/filters.h]: Same as above.

File: bigtable/client/filters.h [hunk 11]
```diff
@@ -214,26 +214,158 @@ class Filter {
...
+    Filter tmp;
+    auto& range = *tmp.filter_.mutable_column_range_filter();
+    range.set_family_name(std::move(column_family));
+    range.set_start_qualifier_open(std::move(start));
+    range.set_end_qualifier_closed(std::move(end));
+    return tmp;
+  }
+
+  /**
+   * Return a filter that accepts columns in the [@p start, @p end] range
```
Comment (mbrukman) [inline bigtable/client/filters.h]: Same as above.

File: bigtable/client/filters.h [hunk 12]
```diff
@@ -214,26 +214,158 @@ class Filter {
...
+    Filter tmp;
+    auto& range = *tmp.filter_.mutable_column_range_filter();
+    range.set_family_name(std::move(column_family));
+    range.set_start_qualifier_closed(std::move(start));
+    range.set_end_qualifier_closed(std::move(end));
+    return tmp;
+  }
+
+  /**
+   * Return a filter that accepts columns in the (@p start, @p end) range
```
Comment (mbrukman) [inline bigtable/client/filters.h]: Same as above.

File: bigtable/client/filters.h [hunk 13]
```diff
@@ -249,6 +381,23 @@ class Filter {
     return tmp;
   }
 
+  /**
+   * Returns a filter that applies a label to each value.
+   *
+   * Each value accepted by previous filters in modified to include the @p
+   * label.
+   *
+   * @note Currently it is not possible to apply more than one label in a
```
Comment (mbrukman) [inline bigtable/client/filters.h]: > Currently, it is ...

File: bigtable/client/filters.h [hunk 14]
```diff
@@ -249,6 +381,23 @@ class Filter {
...
+   *     filter expression, that is, a chain can only contain a single
+   *     ApplyLabelTransformer() filter.  This limitation may be lifted in
+   *     the future.  It is possible to have multiple ApplyLabelTransformer
+   *     filters in a Union() filter, though in this case each copy of a cell
```
Comment (mbrukman) [inline bigtable/client/filters.h]: > , though in this case, each [...]

File: bigtable/client/filters.h [hunk 15]
```diff
@@ -249,6 +381,23 @@ class Filter {
...
+   *     gets a different label.
+   *
+   * @param label the label applied to each cell.  The labels must be at most 15
+   *     characters long, and must match the `[a-z0-9\\-]` pattern.
+   * TODO(#84) - change this if we decide to validate inputs in the client side
```
Comment (mbrukman) [inline bigtable/client/filters.h]: Will this TODO be automatically included in the documentation on the previous line for `@param label`? Did you want that to happen? Do we need a blank line between the arg doc comment and the TODO to dissociate them?

Same comment elsewhere, I've seen this pattern a few times.

File: bigtable/client/filters.h [hunk 16]
```diff
@@ -214,26 +214,158 @@ class Filter {
...
+    return tmp;
+  }
+
+  /**
+   * Return a filter that accepts columns in the (@p start, @p end] range
+   * within the @p column_family.
+   *
+   * TODO(#84) - document what happens if end < start
+   */
+  static Filter ColumnRangeLeftOpen(std::string column_family,
```
Comment (mbrukman) [inline bigtable/client/filters.h]: @sduskis, @garye, @DanielMahu — do we actually support column range queries? I thought we only support row range queries, and column families and column qualifiers need to be specified explicitly, one at a time.

Comment (garye) [reply bigtable/client/filters.h]: ColumnRangeFilters can be applied just like any other filter:
https://github.com/googleapis/googleapis/blob/master/google/bigtable/v2/data.proto#L390

Comment (mbrukman) [reply bigtable/client/filters.h]: OK, got it, so column families are discrete (must be specified explicitly), while column qualifiers are continuous (can be selected individually, or via a range).

File: bigtable/client/filters.h (hunk 1 above)
Comment (coryan) [reply bigtable/client/filters.h]: The rest of the stack expects a `int32`.

File: bigtable/client/filters.h (hunk 2 above)
Comment (coryan) [reply bigtable/client/filters.h]: Ditto.

File: bigtable/client/filters.h (hunk 3 above)
Comment (coryan) [reply bigtable/client/filters.h]: Ditto.

File: bigtable/client/filters.h (hunk 4 above)
Comment (coryan) [reply bigtable/client/filters.h]: Did the first, I like making it explicit that is a floating point range, but will change if you feel strongly about it.

Comment (mbrukman) [reply bigtable/client/filters.h]: Floating point range SGTM.

File: bigtable/client/filters.h (hunk 5 above)
Comment (coryan) [reply bigtable/client/filters.h]: Fixed.

File: bigtable/client/filters.h (hunk 6 above)
Comment (coryan) [reply bigtable/client/filters.h]: Fixed.

File: bigtable/client/filters.h (hunk 7 above)
Comment (coryan) [reply bigtable/client/filters.h]: Fixed.

File: bigtable/client/filters.h (hunk 8 above)
Comment (coryan) [reply bigtable/client/filters.h]: Fixed.

File: bigtable/client/filters.h (hunk 9 above)
Comment (coryan) [reply bigtable/client/filters.h]: Fixed.

File: bigtable/client/filters.h (hunk 10 above)
Comment (coryan) [reply bigtable/client/filters.h]: Fixed.

File: bigtable/client/filters.h (hunk 11 above)
Comment (coryan) [reply bigtable/client/filters.h]: Fixed.

File: bigtable/client/filters.h (hunk 12 above)
Comment (coryan) [reply bigtable/client/filters.h]: Fixed.

File: bigtable/client/filters.h (hunk 13 above)
Comment (coryan) [reply bigtable/client/filters.h]: Fixed.

File: bigtable/client/filters.h (hunk 14 above)
Comment (coryan) [reply bigtable/client/filters.h]: Fixed.

File: bigtable/client/filters.h (hunk 15 above)
Comment (coryan) [reply bigtable/client/filters.h]: Did not want that, thanks for catching it.  Fixed, and looked for the pattern elsewhere too.

**Comment by coryan** [comment]
Ping.  @mbrukman I think I addressed the issues you raised. @garye can you take a look?

File: bigtable/client/filters.h [hunk 17]
```diff
@@ -71,8 +74,12 @@ class Filter {
     return tmp;
   }
 
-  /// Create a filter that accepts only the last @a n values.
-  static Filter Latest(int n) {
+  /**
```
Comment (mbrukman) [inline bigtable/client/filters.h]: Do we have a pattern for when we use `/** ... */` vs. when we use `///` comments? Does Doxygen handle both of them? Do they behave differently somehow?

This should be documented in the style guide, and we should be consistent, assuming they are identical.

File: bigtable/client/filters.h [hunk 18]
```diff
@@ -214,26 +212,158 @@ class Filter {
...
+
+  //@{
+  /**
+   * @name Less common range filters.
+   *
+   * Cloud Bigtable range filters can include or exclude the limits of the
+   * range.  In most cases applications use [start,end) ranges, and the
+   * ValueRange() and ColumnRange() functions are offered to support the
+   * common case.  For the less common cases where the application needs
+   * different ranges the following functions are available.
```
Comment (mbrukman) [inline bigtable/client/filters.h]: s/ranges the/ranges, the/

File: bigtable/client/filters.h [hunk 19]
```diff
@@ -214,26 +212,158 @@ class Filter {
...
+    tmp.filter_.set_row_sample_filter(probability);
+    return tmp;
+  }
+
+  //@{
+  /**
+   * @name Less common range filters.
+   *
+   * Cloud Bigtable range filters can include or exclude the limits of the
+   * range.  In most cases applications use [start,end) ranges, and the
```
Comment (mbrukman) [inline bigtable/client/filters.h]: s/start,end/start, end/   (add space after comma)

File: bigtable/client/filters.h (hunk 17 above)
Comment (coryan) [reply bigtable/client/filters.h]: The style guide proposed in #98 includes guidelines for when to use `/** */` vs. `///` and yes, Doxygen supports both.

File: bigtable/client/filters.h (hunk 19 above)
Comment (coryan) [reply bigtable/client/filters.h]: Fixed.

File: bigtable/client/filters.h (hunk 18 above)
Comment (coryan) [reply bigtable/client/filters.h]: Fixed.

**Pull request merged**